File: python/knapsack.py
class Knapsack:
    """
    Knapsack objects provide an intuitive means of binding
    the problem data 
    """
    
    def __init__(self, cap):
        """
        Initialize new knapsack object
        with capacity and empty items list
        """
        self.capacity = cap
        self.items = []

    def setCapacity(self, cap):
        self.capacity = cap
    
    def addItem(self, weight, value):
        """
        add an item to this knapsack list of items
        """
        self.items.append([weight, value])

    
    def findBest(self, i, j):
        """
        Solve the subproblem of finding i items out of the first j items in the pool
        return the optimal (combined) value
        """

        # check for first cache row
        if i == 0: return 0

        # check for cached result
        if self.cache[i][j] >= 0: return self.cache[i][j]

        # grab item for current row
        weight, val = self.items[i-1]

        # check if the current item exceeds weight limit
        # and adjust cache accordingly
        if weight > j:
            # pass down this call
            new = self.findBest(i-1, j)

            # add result to cache
            self.cache[i][j] = new
            return new
        else:
            # candidates for best value
            bestOld = self.findBest(i-1, j)
            bestNew = self.findBest(i-1, j-weight) + val
            new = max(bestOld, bestNew)

            # add result to cache
            self.cache[i][j] = new
            return new

    
    def solve(self):
        """
        Solve the knapsack problem by finding the item subset with
        maximum cummulative value whose combined weigth is less than
        the knapsack capacity.
        """

        # setup cache
        self.cache = [[-1]*(self.capacity+1) for _ in xrange(len(self.items)+1)]
        self.cache[0] = [0 for _ in xrange(self.capacity+1)]

        # fill cache in a non-recursive way
        for i in xrange(len(self.items)+1):
            self.findBest(i, self.capacity)

        # find optimal subset from cached subproblem solutions
        optimalSubset = []
        col = self.capacity
        for i in xrange(len(self.items), 0, -1):
            if self.cache[i][col] != self.cache[i-1][col]:
                col -= self.items[i-1][0]
                optimalSubset.append(self.items[i-1])

        return self.findBest(len(self.items), self.capacity), optimalSubset

File: boost/src/knapsack.cpp
#include "knapsack.hpp"

// add an item to the pool of items to be considerated for knapsack problem
void Knapsack::addItem(int weight, int value) 
{
	Item nItem;
	nItem.weight = weight;
	nItem.value = value;
	items.push_back(nItem);
}

// Solve the subproblem of finding i items out of the first j items in the pool
// return the optimal (combined) value
int Knapsack::findBest(int i, int j)
{
	// check for first row
	if (i == 0) {return 0;}

	// check for cached result
	if (cache[i][j] >= 0) {return cache[i][j];}

	// grab item for current row
	int weight = items.at(i-1).weight;
	int val = items.at(i-1).value;

	// check if the current item exceeds weight limit
	// and adjust cache accordingly
	if (weight > j)
	{
		// pass down this call
        int n = findBest(i-1, j);

        // add result to cache
        cache[i][j] = n;
        return n;
	}
	// candidates for best value
    int bestOld = findBest(i-1, j);
    int bestNew = findBest(i-1, j-weight) + val;
    int n = bestNew;
    if (bestOld > bestNew) {n = bestOld;}

    // add result to cache
    cache[i][j] = n;
    return n;
}

void Knapsack::setCapacity(int cap) {capacity = cap;}

ItemList Knapsack::solve() 
{
	// init cache
	cache = new int*[items.size()+1];
	cache[0] = new int[capacity+1];
	for (int j=0; j<capacity+1; ++j) {cache[0][j] = 0;}
	for (int i=1; i<items.size()+1; ++i)
	{
		cache[i] = new int[capacity+1];
		for (int j=0; j<capacity+1; ++j) {cache[i][j] = -1;}
	}

	// fill cache in a non-recursive way
	for (int i=0; i<items.size()+1; ++i) {findBest(i, capacity);}

	// find optimal subset from cached subproblem solutions
    ItemList optimalSubset;
    int col = capacity;
    for (int i=items.size(); i>0; --i)
    {
    	if (cache[i][col] != cache[i-1][col])
    	{
    		col -= items.at(i-1).weight;
            optimalSubset.push_back(items.at(i-1));
    	}
    }

	return optimalSubset;
}